core/templates/services/translate.service.ts

```diff
@@ -0,0 +1,177 @@
+  fetchTranslations(lang:string) {
+    return this.http.get(`${this.prefix}${lang}${this.suffix}`).toPromise();
+  }
+
+  // This function sets the new translations
+  use(lang: string) {
+    // Check if it has already been fetched before
+    this.currentLang = lang;
+    if (Object.keys(this.translations).includes(lang)) {
+      this._onLangChange.emit(
+        {lang: lang, translations: this.translations[lang]});
+      return;
+    }
+    // Otherwise fetch the translations
+    this.translations[lang] = this.fetchTranslations(lang).then(
+      translations => {
+        this.translations[lang] = translations;
+        if (this.currentLang === lang) {
+          this._onLangChange.emit(
+            {lang: lang, translations: this.translations[lang]});
+        }
+      }
+    );
+  }
+
+
+  interpolate(expr: string | Function, params?: Object): string {
+    let result: string;
+
+    if (typeof expr === 'string') {
+      result = this.interpolateString(expr, params);
+    } else if (typeof expr === 'function') {
+      result = this.interpolateFunction(expr, params);
+    } else {
+      result = expr as string;
+    }
+
+    return result;
+  }
+
+  getValue(target: Object, key: string): string | undefined {
+    let keys = typeof key === 'string' ? key.split('.') : [key];
+    key = '';
+    let tar: string | undefined;
+    do {
+      key += keys.shift();
+      if (this.utils.isDefined(target) && this.utils.isDefined(target[key]) &&
+        (typeof target[key] === 'object' || !keys.length)) {
+        tar = target[key];
+        target = target[key];
+        key = '';
+      } else if (!keys.length) {
+        target = undefined;
+        tar = undefined;
+      } else {
+        key += '.';
+      }
+    } while (keys.length);
+    return tar;
+  }
+
+  private interpolateFunction(fn: Function, params?: Object | undefined) {
+    return fn(params);
+  }
+
+  private interpolateString(expr: string, params?: Object | undefined) {
+    if (!params) {
+      return expr;
+    }
+
+    return expr.replace(this.templateMatcher,
+      (substring: string, b: string) => {
+        let r = this.getValue(params, b);
+        return this.utils.isDefined(r) ? r : substring;
+      });
+  }
+
+  getParsedResult(
+      translations: Object, key: string, interpolateParams?: Object): string {
+    let res: string;
+
+    if (translations) {
+      res = this.interpolate(
+        this.getValue(translations, key), interpolateParams);
+    }
+
+    // If the translation for the current lang doesn't exist use default lang
+    if (typeof res === 'undefined' && this.defaultLang !== null &&
+      this.defaultLang !== this.currentLang) {
+      res = this.interpolate(this.getValue(
+        this.translations[this.defaultLang], key), interpolateParams);
+    }
+
+    // If the translation for the default lang also doesn't exist raise an error
+    if (typeof res === 'undefined') {
+      let params: MissingTranslationHandlerParams = {
```

Comment:
We should remove this part since we are not using it anywhere. Just making `res = params.key` is sufficient.

Reply:
Done!